#StochHMM - A Flexible hidden Markov model application and C++ library.

##Currently Under Development

StochHMM is an easy to use and flexible hidden Markov model (HMM) application and C++ library. It implements HMMs from a simple user-defined text file. Thereby, allowing researchers to focus on the design and optimization of the model rather than spending precious time implementing HMM algorithms. 

StochHMM implements standard HMM, HMM with duration and in the future will implement hidden semi-Markov model architectures and algorithms. It grants researchers the power to integrate additional datasets in their HMM to improve predictions. Finally, it adapts HMM algorithms to provide stochastic decoding giving researchers the ability to explore and rank sub-optimal predictions.

We are providing StochHMM as a standalone application and C++ library to give researchers the ability to rapidly develop HMMs.

StochHMM is provided as source code and compiles on Windows, Mac OSX, and Linux. We are providing StochHMM under the MIT open source license to increase accessibility and to give researchers the ability to use it in derivative works without restrictions.

StochHMM is developed by Korf Lab, University of California, Davis

Poster presented a the Genome Informatics Conference in Cambridge, UK  September 6-9, 2012.

##Additional References:

1.  Schroeder, D. I., Lott, P., Korf, I. & LaSalle, J. M. Large-scale methylation domains mark a functional subset of neuronally expressed genes. Genome Res 21, 1583–1591 (2011).

2.	Ginno, P. A., Lott, P. L., Christensen, H. C., Korf, I. & Chédin, F. R-loop formation is a distinctive characteristic of unmethylated human CpG island promoters. Mol. Cell 45, 814–825 (2012).

#For Questions, Suggestions, or Assistance, please contact Paul Lott (plott@ucdavis.edu)
